File: src/jastudio/jastudio/language_services/janome_ex/tokenizing/pre_processing_stage/word_info_entry.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots
from note.vocabulary.pos import POS

if TYPE_CHECKING:
    from jastudio.language_services.jamdict_ex.dict_lookup_result import DictLookupResult
    from note.vocabulary.vocabnote import VocabNote


class WordInfoEntry(Slots):
    def __init__(self, word: str, pos: frozenset[str]) -> None:
        self.word: str = word
        self.parts_of_speech: frozenset[str] = pos

    @property
    def is_ichidan(self) -> bool: return POS.ICHIDAN_VERB in self.parts_of_speech

    @property
    def is_godan(self) -> bool: return POS.GODAN_VERB in self.parts_of_speech

    @property
    def is_intransitive(self) -> bool: return POS.INTRANSITIVE in self.parts_of_speech

    @property
    def answer(self) -> str: raise NotImplementedError()


class VocabWordInfoEntry(WordInfoEntry, Slots):
    def __init__(self, word: str, vocab: VocabNote) -> None:
        self.vocab: VocabNote = vocab
        super().__init__(word, vocab.parts_of_speech.get())

    @property
    @override
    def answer(self) -> str: return self.vocab.get_answer()


class DictWordInfoEntry(WordInfoEntry, Slots):
    def __init__(self, word: str, dict_result: DictLookupResult) -> None:
        self.dict_result: DictLookupResult = dict_result
        super().__init__(word, dict_result.parts_of_speech())

    @property
    @override
    def answer(self) -> str: return self.dict_result.format_answer()

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/Tokenizing/PreProcessingStage/WordInfoEntry.cs
using System.Collections.Frozen;
using JAStudio.Core.LanguageServices.JamdictEx;
using JAStudio.Core.Note.Vocabulary;

namespace JAStudio.Core.LanguageServices.JanomeEx.Tokenizing.PreProcessingStage;

public abstract class WordInfoEntry
{
   public string Word { get; }
   public FrozenSet<string> PartsOfSpeech { get; }

   protected WordInfoEntry(string word, FrozenSet<string> partsOfSpeech)
   {
      Word = word;
      PartsOfSpeech = partsOfSpeech;
   }

   public bool IsIchidan => PartsOfSpeech.Contains(POS.IchidanVerb);
   public bool IsGodan => PartsOfSpeech.Contains(POS.GodanVerb);
   public bool IsIntransitive => PartsOfSpeech.Contains(POS.Intransitive);

   public abstract string Answer { get; }
}

class VocabWordInfoEntry : WordInfoEntry
{
   readonly VocabNote _vocab;

   public VocabWordInfoEntry(string word, VocabNote vocab)
      : base(word, vocab.PartsOfSpeech.Get().ToFrozenSet()) =>
      _vocab = vocab;

   public override string Answer => _vocab.GetAnswer();
}

class DictWordInfoEntry : WordInfoEntry
{
   readonly DictLookupResult _dictResult;

   public DictWordInfoEntry(string word, DictLookupResult dictResult)
      : base(word, dictResult.PartsOfSpeech().ToFrozenSet()) =>
      _dictResult = dictResult;

   public override string Answer => _dictResult.FormatAnswer();
}
